Lib/sam.c

```diff
@@ -46,23 +58,322 @@ void ResetHandler();
+#ifdef _SAM4S_HSMCI_INSTANCE_
+        .pfnHSMCI_Handler  = (void*) HSMCI_Handler,  /* 18 Multimedia Card Interface */
+#else
+        .pvReserved18      = (void*) (0UL),          /* 18 Reserved */
+#endif /* _SAM4S_HSMCI_INSTANCE_ */
+        .pfnTWI0_Handler   = (void*) TWI0_Handler,   /* 19 Two Wire Interface 0 */
+        .pfnTWI1_Handler   = (void*) TWI1_Handler,   /* 20 Two Wire Interface 1 */
+        //.pfnSPI_Handler    = (void*) SPI_Handler,    /* 21 Serial Peripheral Interface */ //FIXME
+        .pfnSSC_Handler    = (void*) SSC_Handler,    /* 22 Synchronous Serial Controller */
+        .pfnTC0_Handler    = (void*) TC0_Handler,    /* 23 Timer/Counter 0 */
+        .pfnTC1_Handler    = (void*) TC1_Handler,    /* 24 Timer/Counter 1 */
+        .pfnTC2_Handler    = (void*) TC2_Handler,    /* 25 Timer/Counter 2 */
+#ifdef _SAM4S_TC1_INSTANCE_
+        .pfnTC3_Handler    = (void*) TC3_Handler,    /* 26 Timer/Counter 3 */
+#else
+        .pvReserved26      = (void*) (0UL),          /* 26 Reserved */
+#endif /* _SAM4S_TC1_INSTANCE_ */
+#ifdef _SAM4S_TC1_INSTANCE_
+        .pfnTC4_Handler    = (void*) TC4_Handler,    /* 27 Timer/Counter 4 */
+#else
+        .pvReserved27      = (void*) (0UL),          /* 27 Reserved */
+#endif /* _SAM4S_TC1_INSTANCE_ */
+#ifdef _SAM4S_TC1_INSTANCE_
+        .pfnTC5_Handler    = (void*) TC5_Handler,    /* 28 Timer/Counter 5 */
+#else
+        .pvReserved28      = (void*) (0UL),          /* 28 Reserved */
+#endif /* _SAM4S_TC1_INSTANCE_ */
+        .pfnADC_Handler    = (void*) ADC_Handler,    /* 29 Analog To Digital Converter */
+#ifdef _SAM4S_DACC_INSTANCE_
+        .pfnDACC_Handler   = (void*) DACC_Handler,   /* 30 Digital To Analog Converter */
+#else
+        .pvReserved30      = (void*) (0UL),          /* 30 Reserved */
+#endif /* _SAM4S_DACC_INSTANCE_ */
+        .pfnPWM_Handler    = (void*) PWM_Handler,    /* 31 Pulse Width Modulation */
+        .pfnCRCCU_Handler  = (void*) CRCCU_Handler,  /* 32 CRC Calculation Unit */
+        .pfnACC_Handler    = (void*) ACC_Handler,    /* 33 Analog Comparator */
+        //.pfnUDP_Handler    = (void*) UDP_Handler     /* 34 USB Device Port */ //FIXME
+};
+
+
 // ----- Flash Configuration -----
 
 // ----- Functions -----
 
+void *memset( void *addr, int val, unsigned int len )
+{
+	char *buf = addr;
+
+	for (; len > 0; --len, ++buf)
+		*buf = val;
+	return (addr);
+}
+
+int memcmp( const void *a, const void *b, unsigned int len )
+{
+	const uint8_t *ap = a, *bp = b;
+	int val = 0;
+
+	for (; len > 0 && (val = *ap - *bp) == 0; --len, ++ap, ++bp)
+		/* NOTHING */;
+	return (val);
+}
+
+void *memcpy( void *dst, const void *src, unsigned int len )
+{
+	char *dstbuf = dst;
+	const char *srcbuf = src;
+
+	for (; len > 0; --len, ++dstbuf, ++srcbuf)
+		*dstbuf = *srcbuf;
+	return (dst);
+}
+
+
 // ----- Chip Entry Point -----
 
-// __attribute__ ((section(".startup"))) XXX (HaaTa) Necessary?
 void ResetHandler()
```

Comment:
Where do you set this code to run?

Reply:
With .pfnReset_Handler in the exception_table vectors section (https://github.com/smasher816/controller/blob/a966d74b97ff2f1348ae1554c1e8a8914206c6a5/Lib/sam.c#L151)